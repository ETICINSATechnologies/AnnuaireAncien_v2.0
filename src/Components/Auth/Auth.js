class Auth {
    static connect(jwt){
        sessionStorage.setItem('token', jwt);
    }

    static disconnect(){
        sessionStorage.removeItem('token')
    }

    static isConnected(){
        return !!sessionStorage.getItem('token');
    }

    static getToken() {
        return sessionStorage.getItem('token')
    }

    static getPayload(){
        return sessionStorage.getItem('token');
    }

    static addCorrectButton(buttons){
        if (this.isConnected())
            buttons.push("disconnection");
        else
            buttons.push("connection");
        return buttons;
    }
}

export default Auth;
