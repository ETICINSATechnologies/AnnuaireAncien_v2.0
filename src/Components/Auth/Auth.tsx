class Auth {
    static connect(jwt: string) {
        sessionStorage.setItem('token', jwt);
    }

    static disconnect(): void {
        sessionStorage.removeItem('token')
    }

    static isConnected(): boolean {
        return !!sessionStorage.getItem('token');
    }

    static getToken(): string {
        return sessionStorage.getItem('token') as string;
    }

    static addCorrectButton(buttons: Array<string>) {
        if (this.isConnected())
            buttons.push("disconnection");
        else
            buttons.push("connection");
        return buttons;
    }
}

export default Auth;
