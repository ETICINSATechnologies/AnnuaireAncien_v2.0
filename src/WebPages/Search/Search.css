.Search {
  display: grid;
  height: 100%;
}

.Search .searchForm {
  display: grid;
  grid-template-rows: repeat(5, 4vw) 1fr;
  grid-template-columns: auto 1fr;
  width: 80%;
  border-radius: 1vw;
  border: 0.2vw solid black;
  margin: auto auto 2vw;
  padding-top: 1.5vw;
  padding-bottom: 1.5vw;
  background-color: var(--backgroundSearch);
}

.Search .searchForm * {
  margin: auto;
}

.Search .searchForm div {
  width: auto;
  font-size: 1.3vw;
  margin: auto 1vw auto 1.5vw;
}

.Search .searchForm input[type="text"] {
  width: 85%;
  font-size: 1.2vw;
  padding: 0.5vw;
  margin-left: 0;
}

.Search .searchForm select {
  width: 92%;
  font-size: 1.2vw;
  padding: 0.5vw;
  margin-left: 0;
}

.Search .searchForm .searchInput {
  grid-column: 2;
  font-size: 2vw;
  margin: 1vw 3vw auto auto;
  width: auto;
  background: var(--backgroundButton);
  border: 0.15vw solid black;
}

.Search .searchForm .searchInput:hover {
  cursor: pointer;
}

@media (orientation: landscape) {
  .Search {
    grid-template-columns: 1fr 1fr 1fr;
    overflow: auto;
  }

  .Search caption {
    font-size: 1.5vw;
  }

  .Search tbody {
    max-height: calc(85vh - 6.5vw - 2.5vmax);
  }

  .Search td,
  th {
    width: 6.5vw;
    max-width: 6.5vw;
    font-size: 0.8vw;
    height: 1vw;
  }

  .Search td:last-child {
    width: 2vw;
  }

  .Search th:last-child {
    width: 3.1vw;
  }

  .infoTitle {
    font-size: 1.5vw;
    margin: 1vw 0 1vmin 0;
    text-align: center;
  }

  /*CSS for members array*/
  .Search .members {
    margin: 1vw auto auto;
    max-height: calc(90vh - 6.5vw - 2.5vmax);
  }

  .Search .MemberInfo {
    --default-font-size: 1.5vw;
    --name-font-size: 2.2vw;
    --telephone-font-size: 1.6vw;
    --input-label-font-size: 1.1vw;

    max-height: calc(90vh - 6.5vw - 2.5vmax);
  }

  .Search .MemberInfo .infoArea {
    width: 80%;
    border: 0.2vw solid black;
    border-radius: 1vw;
    margin: auto auto 2vw;
  }
}

@media (orientation: portrait) {
  .Search {
    grid-template-columns: 1fr 1fr;
    grid-template-rows: 2fr 10fr;
  }

  .Search .searchArea {
    width: 90%;
    margin: 3vw auto auto;
  }

  /*CSS for members array*/
  .Search .members {
    grid-row: 2;
    grid-column: 1 / 3;
    margin: 3vmin auto auto;
  }

  .Search caption {
    font-size: 3vw;
  }

  .Search .members tbody {
    max-height: calc(90vh - 6.5vw - 2.5vmax - 65vw);
  }

  .Search .members td,
  .Search .members th {
    width: 24vw;
    max-width: 24vw;
    font-size: 2vw;
    height: 2.4vw;
  }

  .Search .members td:last-child {
    width: 6vw;
  }

  .Search .members th:last-child {
    width: auto;
  }

  /*CSS for Board array*/
  .Search .CA {
    grid-row: 1;
    grid-column: 1;
    margin: 3vw auto auto;
  }

  .Search .CA tbody {
    max-height: 27vh;
  }

  .Search .CA td,
  .Search .CA th {
    width: 8.5vw;
    max-width: 8.5vw;
    font-size: 1vw;
    height: 1.2vw;
  }

  .Search .CA td:last-child {
    width: 3vw;
  }

  .Search .CA th:last-child {
    width: 3vw;
  }

  .Search .MemberInfo {
    --default-font-size: 2.2vw;
    --name-font-size: 3vw;
    --telephone-font-size: 2.5vw;
    --input-label-font-size: 2vw;

    width: 95%;
    max-height: calc(90vh - 6.5vw - 2.5vmax);
    margin: 3vw auto auto;
  }

  .Search .infoTitle {
    font-size: 3vw;
    margin: 0 0 1vmin 0;
    text-align: center;
  }

  .Search .searchForm {
  }

  /*CSS for member info*/
  .Search .MemberInfo .infoArea {
    grid-template-rows: 2vw 4.5vw 4.5vw 4vw repeat(5, 2.5vw 5.1vw);
    width: 80%;
    border: 0.3vmin solid black;
    margin: auto 13% auto auto;
  }
}
