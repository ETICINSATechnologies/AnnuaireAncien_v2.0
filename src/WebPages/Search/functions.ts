import Auth from "../../Components/Auth/Auth";
import { parse } from "json2csv";

// To-do : Needs more complex page handling. Right now, will only download first page
export const exportAllAlumni = async () => {
    let pageSize = 25;
    let totalPages = 1;

    let alumniList: any[] = [];

    for (let pageNumber = 0; pageNumber < totalPages; ++pageNumber) {
        try {
            const results = await fetchResults(pageSize, pageNumber);
            if (results && results.meta && results.content) {
                totalPages = results.meta.totalPages;
                alumniList = [...alumniList, ...results.content];
            }
        } catch (error) {
            alert(error);
            return;
        }
    }

    const flattenedAlumniList = alumniList.map((item: any) => flattenObject(item));
    try {
        const csv = 'data:text/csv;charset=utf-8,' + parse(flattenedAlumniList);
        const encodedUri = encodeURI(csv);
        const link = document.createElement("a");
        link.setAttribute("href", encodedUri);
        link.setAttribute("download", "alumni_etic.csv");
        document.body.appendChild(link); // Required for FF
        link.click();
    } catch (error) {
        alert(error);
        return;
    }
}

const fetchResults = async (pageSize: number, pageNumber: Number) => {
    const url: string = `${process.env.REACT_APP_BACKEND_URL}/api/v1/core/member?isAlumni=true&pageSize=${pageSize}&pageNumber=${pageNumber}`;
    const response = await fetch(url, {
        headers: {
            Authorization: Auth.getToken()
        }
    });
    if (response.ok) {
        const results = await response.json();
        return results;
    } else {
        throw new Error("Error fetch members : " + response.statusText);
    }

}

const flattenObject = (obj: any, parent?: any, res = {} as any): Object => {
    for (let key in obj) {
        let propName = parent ? parent + '_' + key : key;
        if (typeof obj[key] == 'object') {
            flattenObject(obj[key], propName, res);
        } else if (Array.isArray(obj[key])) {
            res[propName] = flattenArray(obj[key]);
        } else {
            res[propName] = obj[key];
        }
    }
    return res;
};

const flattenArray = (arr: Array<any>): Array<any> => {
    let flattened = arr.flat(Infinity);
    return flattened.map((element) => {
        if (typeof element == 'object') {
            return flattenObject(element);
        } else {
            return element;
        }
    });
};