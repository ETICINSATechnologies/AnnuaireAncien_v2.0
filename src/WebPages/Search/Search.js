import React, {Component} from 'react';
import './Search.css';
import Header from "../../Components/Header/Header";
import Nav from "../../Components/Nav/Nav";
import Auth from "../../Components/Auth/Auth";
import {Redirect} from 'react-router-dom'
import MemberArray from "../../Components/MemberArray/MemberArray";
import MemberInfo from "../../Components/MemberInfo/MemberInfo";


import Redirecting from "../Redirecting.js";

class Search extends Component {
    constructor(props) {
        super(props);
        this.searchFields = {
            firstName: 'Prénom',
            lastName: 'Nom',
            positionId: 'Poste',
            company: 'Entreprise',
            year: 'Année'
        };
        this.state = {
            status: 'noParameters', // 'PARAMETETERS' 'NO_PARAMETERS' 'pending'
            memberInfo: {},
        };
        this.selectMember = this.selectMember.bind(this);
        this.makeSearch = this.makeSearch.bind(this)
    }

    selectMember(member) {
        this.setState({memberInfo: member});
    }

    makeSearch(event) {
        event.preventDefault();

        this.setState({status: 'noParameters'});

        let searchArray = {};
        Object.keys(this.searchFields).forEach((fieldName) => {
            let val;
            if (val = document.getElementById(`${fieldName}Input`).value) {
                searchArray[fieldName] = val;
                this.setState({status: 'Parameters'});
            }
        });

        this.refs.members.getMembers(searchArray);
    }

    render() {

        let activeButton = ["home"];
        if (!Auth.isConnected())
            return <Redirect to='/'/>;

        activeButton.push('profile');
        activeButton = Auth.addCorrectButton(activeButton);
        let renderSearchFields = [];
        Object.keys(this.searchFields).forEach((fieldName) => {
            renderSearchFields.push(
                <div className={`${fieldName}Div`} key={`${fieldName}Div`}> {this.searchFields[fieldName]} </div>
            );

            renderSearchFields.push(
                <input type="text" id={`${fieldName}Input`} className={`${fieldName}Input`} key={`${fieldName}Input`}/>
            );

        });


        return (
            <React.Fragment>
                <Header/>

                <Nav buttons={activeButton}> </Nav>
                <section className="Search">
                    <div className="searchArea">
                        <div className='infoTitle'> Rechercher</div>
                        <form className="searchForm">
                            {renderSearchFields}
                            <input className="searchInput" type="submit" value="Rechercher" onClick={this.makeSearch}/>
                        </form>
                    </div>
                    <MemberArray className="members" onClick={this.selectMember} search ref="members"/>
                    <MemberInfo info={this.state.memberInfo}/>
                </section>

            </React.Fragment>
        )
            ;
    }
}

export default Search;
