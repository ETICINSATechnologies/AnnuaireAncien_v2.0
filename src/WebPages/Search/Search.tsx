import React, {Component} from 'react';
import {Redirect} from "react-router-dom";

import './Search.css';
import Header from "../../Components/Header/Header";
import Nav from "../../Components/Nav/Nav";
import Auth from "../../Components/Auth/Auth";
import MemberArray from "../../Components/MemberArray/MemberArray";
import MemberInfo from "../../Components/MemberInfo/MemberInfo";

import {Position} from "../../Model/Position";
import {Member} from "../../Model/Member";
import {SearchInterface} from "../../Model/Searchinterface";

let defaultMan = require('../../Images/default_man.svg');
let defaultWoman = require('../../Images/default_woman.svg');


interface SearchState {
    searchValues: SearchInterface
    positions: Position[]
    selectedMember?: Member
    img: string
}


const searchFields = {
    firstName: 'Prénom',
    lastName: 'Nom',
    positionId: 'Poste',
    company: 'Entreprise',
    year: 'Année',
} as SearchInterface;


class Search extends Component<{}, SearchState> {
    state = {
        searchValues: {
            firstName: '',
            lastName: '',
            positionId: '',
            company: '',
            year: ''
        } as SearchInterface,
        positions: [],
        selectedMember: undefined,
        img: defaultMan
    };

    componentDidMount() {
        fetch('position', {
            headers: {
                Authorization: Auth.getToken()
            }
        })
            .then(res => res.json())
            .then((result) => {
                this.setState({positions: result as Position[]});
            })
    }

    selectMember = (member: Member) => {
        this.setState({
            selectedMember: member
        });
        this.getMemberImage(member);
    };

    getMemberImage = (member: Member) => {
        fetch(`member/${member.id}/image`, {
            headers: {
                'Authorization': Auth.getToken(),
                'Content-Type': 'application/json'
            }
        })
            .then(res => {
                if (res.status === 200) {
                    res.blob()
                        .then(image => {
                            this.setState({
                                img: URL.createObjectURL(image),
                            })
                        })
                } else {
                    this.setState({
                        img: (this.state.selectedMember as any).gender === 'F' ? defaultWoman : defaultMan
                    });
                }
            })
    };

    makeSearch = (page :number,event?: React.MouseEvent) => {
        event? event.preventDefault():null;

        // copy the input into a new variable
        let searchArray = Object.assign({}, this.state.searchValues);

        // remove the empty param fields
        Object.keys(searchArray).forEach((paramName) => {
            if (searchArray[paramName] === "")
                delete searchArray[paramName];
        });

        (this.refs.members as MemberArray).getMembers(searchArray, page);
    };

    updateParameters = (event: React.ChangeEvent) => {
        const paramName = event.target.className;
        let paramValue = (event.target as any).value;

        // update if this is not the year or if the year is a 4 digit number
        if (paramName !== "year" || paramValue.match(/^(\d?){4}$/))
            this.setState({
                searchValues: {
                    ...this.state.searchValues,
                    [event.target.className]: (paramValue)
                }
            });
    };

    render() {
        let activeButton = ["home"];
        if (!Auth.isConnected())
            return <Redirect to='/'/>;

        if(Auth.isAdmin()) {
            activeButton.push('member_creation');
            activeButton.push('data');
        }else{
            activeButton.push('profile');
        }
        activeButton = Auth.addCorrectButton(activeButton);

        // create the dropdown for the positions
        let positionDropDown = this.state.positions.map((position: Position, index) => {
            return <option key={index} value={position.id}>{position.label}</option>;
        });

        // create an array which contains the div and input for each field of the research
        let renderSearchFields: any = [];
        Object.keys(searchFields).forEach((fieldName) => {
            renderSearchFields.push(
                <div className={`${fieldName}Div`} key={`${fieldName}Div`}> {searchFields[fieldName]} </div>
            );
            if (fieldName === 'positionId') {
                renderSearchFields.push(
                    <select key='positionId' className='positionId' onChange={this.updateParameters}>
                        <option value=""/>
                        {positionDropDown}
                    </select>
                )
            }
            else {
                renderSearchFields.push(
                    <input type="text" className={fieldName} key={`${fieldName}Input`}
                           onChange={this.updateParameters} value={this.state.searchValues[fieldName]}/>
                );
            }
        });

        return (
            <React.Fragment>
                <Header/>
                <Nav buttons={activeButton}> </Nav>
                <section className="Search">
                    <div className="searchArea">
                        <div className='infoTitle'> Rechercher</div>
                        <form className="searchForm">
                            {renderSearchFields}
                            <input className="searchInput" type="submit"
                                   value="Rechercher" onClick={(e: React.MouseEvent) => this.makeSearch(0, e)}
                            />
                        </form>
                    </div>
                    <MemberArray parameters={this.state.searchValues} selectMember={this.selectMember} ref="members"
                                 getPage={this.makeSearch}/>
                    <MemberInfo member={this.state.selectedMember} image={this.state.img}/>
                </section>
            </React.Fragment>
        );
    }
}

export default Search;
