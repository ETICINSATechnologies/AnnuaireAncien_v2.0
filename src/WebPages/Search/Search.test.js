import React from 'react';
import ReactDOM from 'react-dom';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import {BrowserRouter as Router} from "react-router-dom";

import Search from './Search';
import Auth from '../../Components/Auth/Auth';

Enzyme.configure({adapter: new Adapter()});

describe('Search component', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(shallow(
            <Router>
                <Search/>
            </Router>), div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('Search connected unit test', () => {
        Auth.isConnected = jest.fn(() => true);
        expect(
            shallow(<Search/>)
                .find('section.Search')
                .length
        ).toBe(1)
    });

    it('Search not connected snapshot test', () => {
        Auth.isConnected = jest.fn(() => false);
        const component = renderer.create(
            <Router>
                <Search/>
            </Router>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

    });

    it('Search connected snapshot test', () => {
        Auth.isConnected = jest.fn(() => true);
        const component = renderer.create(
            <Router>
                <Search/>
            </Router>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })

});