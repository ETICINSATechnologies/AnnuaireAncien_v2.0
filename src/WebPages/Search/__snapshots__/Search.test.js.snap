// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component Search connected snapshot test 1`] = `
Array [
  <header
    className="Header"
  >
    <a
      className="header_a"
      href="/"
    >
      <img
        alt=""
        className="home_logo"
        src="home_logo.png"
      />
      <h1
        className="title"
      >
         Annuaire des Anciens
      </h1>
    </a>
  </header>,
  <nav
    className="Nav"
  >
    <div
      className="Menu"
    >
      <a
        className="button"
        href="/"
      >
        accueil
      </a>
      <a
        className="button"
        href="/profile"
      >
        mon profil
      </a>
      <a
        className="button"
        href="/"
        onClick={[Function]}
      >
        se déconnecter
      </a>
    </div>
    <a
      className="help_link"
      href="/help"
    >
      <img
        alt=""
        className="help_icon"
        src="help_icon.png"
      />
    </a>
  </nav>,
  <section
    className="Search"
  >
    <div
      className="searchArea"
    >
      <div
        className="infoTitle"
      >
         Rechercher
      </div>
      <form
        className="searchForm"
      >
        <div
          className="firstNameDiv"
        >
           
          Prénom
           
        </div>
        <input
          className="firstNameInput"
          id="firstNameInput"
          type="text"
        />
        <div
          className="lastNameDiv"
        >
           
          Nom
           
        </div>
        <input
          className="lastNameInput"
          id="lastNameInput"
          type="text"
        />
        <div
          className="positionIdDiv"
        >
           
          Poste
           
        </div>
        <input
          className="positionIdInput"
          id="positionIdInput"
          type="text"
        />
        <div
          className="companyDiv"
        >
           
          Entreprise
           
        </div>
        <input
          className="companyInput"
          id="companyInput"
          type="text"
        />
        <div
          className="yearDiv"
        >
           
          Année
           
        </div>
        <input
          className="yearInput"
          id="yearInput"
          type="text"
        />
        <input
          className="searchInput"
          onClick={[Function]}
          type="submit"
          value="Rechercher"
        />
      </form>
    </div>
    <section
      className="MemberArray members"
    >
      <table>
        <caption>
          Liste des membres 
          
        </caption>
        <thead>
          <tr
            className="MemberDisplay undefined"
            onClick={[Function]}
          >
            <th
              className="firstName"
            >
              prénom
            </th>
            <th
              className="lastName"
            >
              nom
            </th>
            <th
              className="positionLabel"
            >
              poste
            </th>
            <th
              className="positionYear"
            >
              année
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td />
            <td />
            <td />
            <td />
          </tr>
          <tr>
            <td />
            <td />
            <td />
            <td />
          </tr>
          <tr>
            <td />
            <td />
            <td />
            <td />
          </tr>
          <tr>
            <td />
            <td />
            <td />
            <td />
          </tr>
          <tr>
            <td />
            <td />
            <td />
            <td />
          </tr>
          <tr>
            <td />
            <td />
            <td />
            <td />
          </tr>
          <tr>
            <td />
            <td />
            <td />
            <td />
          </tr>
          <tr>
            <td />
            <td />
            <td />
            <td />
          </tr>
          <tr>
            <td />
            <td />
            <td />
            <td />
          </tr>
          <tr>
            <td />
            <td />
            <td />
            <td />
          </tr>
        </tbody>
      </table>
    </section>
    <section
      className="MemberInfo"
    >
      <div
        className="infoTitle"
      >
         Info du membre
      </div>
      
    </section>
  </section>,
]
`;

exports[`Search component Search not connected snapshot test 1`] = `null`;
